crates/sparrow-session/src/execution.rs

```diff
@@ -0,0 +1,104 @@
+use arrow_array::RecordBatch;
+use futures::future::BoxFuture;
+use futures::stream::BoxStream;
+use futures::{StreamExt, TryStreamExt};
+use sparrow_api::kaskada::v1alpha::ExecuteResponse;
+
+use crate::Error;
+
+pub struct Execution {
+    /// Tokio runtme managing this execution.
+    rt: tokio::runtime::Runtime,
+    /// Channel to receive output on.
+    output: tokio_stream::wrappers::ReceiverStream<RecordBatch>,
+    /// Future which resolves to the first error or  None.
+    status: Status,
+}
+
+enum Status {
+    Running(BoxFuture<'static, error_stack::Result<(), Error>>),
+    Failed,
+    Completed,
+}
+
+impl Execution {
+    pub(super) fn new(
+        rt: tokio::runtime::Runtime,
+        output_rx: tokio::sync::mpsc::Receiver<RecordBatch>,
+        progress: BoxStream<'static, error_stack::Result<ExecuteResponse, Error>>,
+    ) -> Self {
+        let output = tokio_stream::wrappers::ReceiverStream::new(output_rx);
+        let status = Status::Running(Box::pin(async move {
+            let mut progress = progress;
+            while (progress.try_next().await?).is_some() {}
+            Ok(())
+        }));
+
+        Self { rt, output, status }
+    }
+
+    /// Check the status future.
+    ///
+    /// If it has previously completed (successfully or with error) returns
+    /// accordingly. Otherwise, check to see if the future is ready, and update
+    /// status (and return) accordingly.
+    fn is_done(&mut self) -> error_stack::Result<(), Error> {
+        let result = match &mut self.status {
+            Status::Running(future) => {
+                // Based on the implementation of `FutureExt::now_or_never`:
+                let noop_waker = futures::task::noop_waker();
+                let mut cx = std::task::Context::from_waker(&noop_waker);
+
+                match future.as_mut().poll(&mut cx) {
+                    std::task::Poll::Ready(x) => x,
+                    _ => return Ok(()),
+                }
+            }
+            Status::Failed => error_stack::bail!(Error::ExecutionFailed),
+            Status::Completed => return Ok(()),
+        };
+
+        match result {
+            Ok(_) => {
+                self.status = Status::Completed;
+                Ok(())
+            }
+            Err(e) => {
+                self.status = Status::Failed;
+                Err(e)
+            }
+        }
+    }
+
+    pub async fn next(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
+        self.is_done()?;
+        Ok(self.output.next().await)
+    }
+
+    pub async fn next_blocking(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
```

Comment:
`collect_all_blocking` isn't `async` but this is - intentional?

Reply:
Oh. This should be non-async (it uses the tokio `runtime` directly, and waits for that).